### PR TITLE
add scanMnay for Unfold

### PR DESCRIPTION
**Pull request by rnjtranjan**

### CODE REVIEW

**Comment by rnjtranjan** [comment]
Fixes: #1308

File: core/src/Streamly/Internal/Data/Unfold.hs [hunk 1]
```diff
@@ -458,6 +462,47 @@ scan (Fold stepF initial extract) (Unfold stepU injectU) =
...
+                    FL.Done v -> do
+                        if isNothing nxt
+                        then return $ Yield v (Just (fs, s), Just v)
+                        else do
+                            r <- initial
+                            return $ case r of
+                                FL.Partial fs1 ->
+                                    Skip (Just (fs1, us), Nothing)
+                                FL.Done _ ->
+                                    Skip (Nothing, Nothing)
```
Comment (harendra-kumar) [inline core/src/Streamly/Internal/Data/Unfold.hs:496]: This is unnecessarily complicated, why not just say return `Stop` here? The `Maybe` is not required on the first element of the tuple.

Comment (harendra-kumar) [reply core/src/Streamly/Internal/Data/Unfold.hs:496]: Better would be to just change `scan` to `scanWith` and pass a flag to it to stop or restart the scan. Then use the same common function of scan and scanMany.

**Review by harendra-kumar — CHANGES_REQUESTED** [review]

File: core/src/Streamly/Internal/Data/Unfold.hs [hunk 2]
```diff
@@ -475,6 +479,7 @@ either (Unfold step1 inject1) = Unfold step inject
             Skip s -> Skip (s, f)
             Stop -> Stop) <$> step1 st
 
+
```
Comment (harendra-kumar) [inline core/src/Streamly/Internal/Data/Unfold.hs:482]: Remove blank line.

File: core/src/Streamly/Internal/Data/Unfold.hs [hunk 3]
```diff
@@ -507,6 +512,64 @@ postscan (Fold stepF initial extract) (Unfold stepU injectU) =
...
+    where
+
+    inject a =  do
+        r <- initial
+        case r of
+            FL.Partial fs -> do
+                x <- injectU a
+                return (Just (fs, x), Nothing)
+            FL.Done _ -> return (Nothing, Nothing)
```
Comment (harendra-kumar) [inline core/src/Streamly/Internal/Data/Unfold.hs:528]: This is not correct, we need to yield the result before we stop.

**Review by harendra-kumar — CHANGES_REQUESTED** [review]

File: core/src/Streamly/Internal/Data/Unfold.hs [hunk 4]
```diff
@@ -507,6 +511,66 @@ postscan (Fold stepF initial extract) (Unfold stepU injectU) =
...
+    Unfold step inject
+
+    where
+
+    inject a =  do
+        r <- initial
+        case r of
+            FL.Partial fs -> do
+                x <- injectU a
+                return (Just (fs, x), Nothing)
```
Comment (harendra-kumar) [inline core/src/Streamly/Internal/Data/Unfold.hs:526]: A scan always yields the initial value. This is not correct. Can you write a test for this?

**Review by harendra-kumar — CHANGES_REQUESTED** [review]

**Pull request merged**